### PR TITLE
Implement Decoder and Encoder

### LINKED ISSUE #4: Support NewDecoder?
cc https://github.com/ghodss/yaml/issues/51

### PR DESCRIPTION
**Pull request by SVilgelm**
add support for streaming encoding and decoding

closes #4

### CODE REVIEW

**Comment by SVilgelm** [comment]
I have also raise the PR with similar code to `sigs.k8s.io/yaml` fork: https://github.com/kubernetes-sigs/yaml/pull/79

**Review by samlown — COMMENTED** [review]
I like the look of where this is going! Thanks for submitting! We've already started using the Decoder and Encoder methods, so perhaps it would be easier to enabled wrappers around those? I see a few copy and pasted comments, which suggests to me "refactor".

File: yaml.go [hunk 1]
```diff
@@ -307,3 +307,125 @@ func convertToJSONableObject(yamlObj interface{}, jsonTarget *reflect.Value) (in
...
+	if err := jsonUnmarshal(&buf, obj, dec.opts...); err != nil {
+		return fmt.Errorf("error unmarshaling JSON: %w", err)
+	}
+
+	return nil
+}
+
+// KnownFields ensures that the keys in decoded mappings to
+// exist as fields in the struct being decoded into.
+func (dec *Decoder) KnownFields() {
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply yaml.go:421]: sure, I will add the test for this method

File: yaml.go [hunk 2]
```diff
@@ -307,3 +307,125 @@ func convertToJSONableObject(yamlObj interface{}, jsonTarget *reflect.Value) (in
...
+	dec.decoder.KnownFields(true)
+	dec.opts = append(dec.opts, DisallowUnknownFields)
+}
+
+// DisallowUnknownFields configures the JSON decoder to error out if unknown
+// fields come along, instead of dropping them by default.
+func DisallowUnknownFields(d *json.Decoder) *json.Decoder {
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply yaml.go:428]: I think it is better to have it as a separate JSONOpt, it is in original ghodss code and we use it a lot, but on another hand it should be always used by calling KnownFields method. will refactor

**Comment by SVilgelm** [comment]
> it would be easier to enabled wrappers around those

as I said I just copied my PR I raised for sigs.k8s.io fork and did not fully adapt it to your code, let me see how I can make it better

**Comment by SVilgelm** [comment]
@samlown Hi, I have refactored Encoder/Decoder to reuse the existent code, extend the tests for testing Unmarshal/Decode and Marshal/Encode functions with strict and non strict mode

File: yaml.go [hunk 3]
```diff
@@ -70,27 +143,32 @@ func jsonUnmarshal(r io.Reader, o interface{}, opts ...JSONOpt) error {
...
-		return fmt.Errorf("while decoding JSON: %v", err)
+		return fmt.Errorf("while decoding JSON: %w", err)
 	}
 	return nil
 }
 
 // JSONToYAML converts JSON to YAML.
 func JSONToYAML(j []byte) ([]byte, error) {
+	var buf bytes.Buffer
+	err := jsonToYAML(yaml.NewEncoder(&buf), bytes.NewReader(j))
```
Comment (samlown) [inline yaml.go:154]: Again, the encoder might need closing.

File: yaml_test.go [hunk 4]
```diff
@@ -145,69 +161,107 @@ func TestUnmarshalNonStrict(t *testing.T) {
...
-// prettyFunctionName converts a slice of JSONOpt function pointers to a human
-// readable string representation.
-func prettyFunctionName(opts []JSONOpt) []string {
-	var r []string
-	for _, o := range opts {
-		r = append(r, runtime.FuncForPC(reflect.ValueOf(o).Pointer()).Name())
-	}
-	return r
-}
+func unmarshalEqual(t *testing.T, y []byte, s, e interface{}, knowFields bool) {
```
Comment (samlown) [inline yaml_test.go:168]: This should probably be `knownFields`. I'm not a fan of boolean flags though, they're hard to read. Might it be better to define specific functions for each use-case? i.e.:

* `unmarshalEqual`
* `decodeEqual`
* `decodeEqualKnownFields`

**Review by samlown — COMMENTED** [review]
Looking good. I'd just check if `Close` is required in the Encoders, and double check the `io.EOF` issue.

**Comment by SVilgelm** [comment]
thank for reviewing, I agree that we need to close the decoder even if it is not required, give me a day or tow and I will fix it

**Pull request closed without merging**